### PR TITLE
Port select-only type validation logic to builder

### PR DESCRIPTION
**Pull request by Eiji7**
solves #2496

@josevalim I was not sure how to write tests well, so I created scenario for my case as described in issue, so I'm casting type of json value in `order_by` macro for both adapters.

Hope I did not missed anything.

### CODE REVIEW

File: test/ecto/adapters/mysql_test.exs [hunk 1]
```diff
@@ -44,6 +44,16 @@ defmodule Ecto.Adapters.MySQLTest do
     end
   end
 
+  defmodule Schema4 do
```
Comment (josevalim) [inline test/ecto/adapters/mysql_test.exs]: Please don't add a new schema. Just use one of the existing ones.

File: test/ecto/adapters/postgres_test.exs [hunk 2]
```diff
@@ -46,6 +46,16 @@ defmodule Ecto.Adapters.PostgresTest do
     end
   end
 
+  defmodule Schema4 do
```
Comment (josevalim) [inline test/ecto/adapters/postgres_test.exs]: Please don't add a new schema, just use one of the existing ones.

File: test/ecto/adapters/postgres_test.exs [hunk 3]
```diff
@@ -395,6 +405,19 @@ defmodule Ecto.Adapters.PostgresTest do
...
+    query =
+      normalize from(e in "schema4",
+        order_by: type(fragment("(?->>?)", field(e, ^:map), ^"value"), ^:decimal),
+        select: true)
+
+    result =
+      "SELECT TRUE FROM \"schema4\" AS s0 " <>
+      "ORDER BY (s0.\"map\"->>$1)::decimal"
+
+    assert all(query) == String.trim(result)
```
Comment (josevalim) [inline test/ecto/adapters/postgres_test.exs]: You don't need trim here. YOu can keep everything in a single line:

    assert all(query) ==  "SELECT TRUE FROM \"schema4\" AS s0 ORDER BY (s0.\"map\"->>$1)::decimal"

File: test/ecto/adapters/mysql_test.exs [hunk 4]
```diff
@@ -222,6 +232,19 @@ defmodule Ecto.Adapters.MySQLTest do
...
+    query =
+      normalize from(e in "schema4",
+        order_by: type(fragment("(?->>?)", field(e, ^:map), ^"value"), ^:decimal),
+        select: true)
+
+    result =
+      "SELECT TRUE FROM `schema4` AS s0 " <>
+      "ORDER BY (s0.`map`->>?) + 0"
+
+    assert all(query) == String.trim(result)
```
Comment (josevalim) [inline test/ecto/adapters/mysql_test.exs]: You don't need trim here. YOu can keep everything in a single line:

    assert all(query) ==  "SELECT TRUE FROM \"schema4\" AS s0 ORDER BY (s0.\"map\"->>$1)::decimal"

**Comment by josevalim** [comment]
@Eiji7 i have dropped some comments and this should be good to go!

**Comment by Eiji7** [comment]
@josevalim Is it ok now?

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**